gatsbyjs/gatsby
# feat(gatsby-plugin-offline): replace no-cache detection with dynamic path whitelist

Fixes #9355, fixes #9415, fixes #9661, fixes #9939

Changes made:

- `gatsby-plugin-offline`
  - `gatsby-browser.js`
    - Detect if the SW is installed with a more reliable method
    - Whitelist pathnames which have been prefetched
  - `gatsby-node.js`
    - Remove navigate fallback options
    - Move the catch-all Network First handler to `sw-append.js`
    - Copy `idb-keyval-iife.min.js` to the `public` folder when building
  - `sw-append.js`
    - Define a custom navigation route, which serves the offline shell for whitelisted paths
    - Refactor the API to separate individual functions from the message handler
    - Add new APIs to whitelist a path, and to reset the whitelist
  - `README.md`
    - Update default options
    - Remove reference to AppCache (replaced by service workers)
- `gatsby/cache-dir`
  - `ensure-resources.js`
    - Remove all `load-directly-or-404.js` code
    - Upon a resource load failure, reload the page in order to display a server 404 or offline error
  - `load-directly-or-404.js` - delete it
  - `loader.js`
    - Remove unnecessary `stripPrefix` code (the prefix is already stripped so it caused errors)
    - Remove unnecessary `failedResources` object (its values are never read)
    - Delete promises from `jsonPromiseStore` if the resources fail to load
    - Show errors from `fetchResource` in the console, rather than silently catching
    - Use object shorthand notation where possible
    - Only call `onPostPrefetchPathname` from `getResourcesForPathname` and `enqueue` if they were successful (as per the docs, it should only be called upon success)
  - `prefetch.js`
    - Detect whether prefetching was successful
  - `navigation.js`
    - Remove all `load-directly-or-404.js` code
    - Reset the IDB whitelist if the service worker has updated
  - `production-app.js`
    - Remove all `load-directly-or-404.js` code

TODO:

- [x] Figure out the cause of #9415
- [x] ...and fix it + port the fix to its own PR
- [x] Check offline plugin README matches latest default config
- [x] Check if `onPostPrefetchResources` can be called from `enqueue`
- [x] Describe exactly what I've changed
- [x] Check all manual tests pass
- [x] Remove excessive debugging statements
- [x] Bundle idbKeyval rather than fetching it from the internet
- ~~Fix resources failing to load after going offline, then coming online again and hovering over links~~ see webpack-contrib/mini-css-extract-plugin/issues/278

---

Manual tests:

```
configurations:

- 1) online without gatsby-plugin-offline
- 2) online with gatsby-plugin-offline
- 3) offline with gatsby-plugin-offline (start on /about to prevent all URLs prefetching)
- 4) online in develop

- x.1) with a custom 404 page
- x.2) without a custom 404 page

tests:
- 1) 404s are handled correctly from internal links - custom 404 or server 404
  if custom is unavailable, or offline error, but no blank pages
- 2) 404s are handled correctly when entering a URL directly - as above, and
  the URL fetched contains the original query parameters (not just no-cache=1)
- 3A) back button works properly after internal link 404 (navigates to the
  expected page)
- 3B) back button works properly after external link 404 (navigates to the
  expected page)
- 4) pages which fail to load display a native offline error (no blank pages)
- 5) the service worker remains installed after a 404/offline error
- 6) Netlify CMS /admin/ loads when reloaded / URL directly entered (or offline
  error displayed)
- 7) going to a page by entering the URL directly works
- 8) directly entering the mismatching canonical URL works with JS, and goes
  backwards and forwards correctly (or shows offline)
```
